### ISSUE TITLE
Error":"missing required startup taint, karpenter.sh/unregistered" in the karpenter logs.

### ISSUE THREAD

**Issue by janani2019** [kubernetes-sigs/karpenter#1635]
### Description

**Observed Behavior**:
I have startup taints in my nodepool but I still see this error
{"level":"ERROR","time":"2024-09-04T20:20:36.965Z","logger":"controller","message":"Reconciler error","commit":"62a726c","controller":"nodeclaim.lifecycle","controllerGroup":"karpenter.sh","controllerKind":"NodeClaim","NodeClaim":{"name":"deployment-rqczg"},"namespace":"","name":"deployment-rqczg","reconcileID":"db521016-5e44-4c77-9352-f07bbb44197d","error":"missing required startup taint, karpenter.sh/unregistered"}
**Expected Behavior**:
This error should not coming up and the start up taints should be applied.
**Reproduction Steps** (Please include YAML):
apiVersion: karpenter.sh/v1
kind: NodePool
metadata:
  name: deployment
spec:
  template:
    metadata:
      labels:
        type: DeploymentRunners
    spec:
      expireAfter: 19m
      terminationGracePeriod: 20m
      requirements:
      - key: kubernetes.io/arch
        operator: In
        values: ["amd64"]
      - key: kubernetes.io/os
        operator: In
        values: ["linux"]
      - key: karpenter.sh/capacity-type
        operator: In
        values: ["on-demand"]
      - key: karpenter.k8s.aws/instance-category
        operator: In
        values: ["c", "m", "r"]
      - key: karpenter.k8s.aws/instance-generation
        operator: Gt
        values: ["2"]
      - key: node.kubernetes.io/instance-type
        operator: Exists
        minValues: 10
      nodeClassRef:
        group: karpenter.k8s.aws
        kind: EC2NodeClass
        name: default
      taints:
      - key: "deployment_runners_workload"
        value: "true"
        effect: NoSchedule
      startupTaints:
      - key: karpenter.sh/unregistered
        effect: NoSchedule
  limits:
    cpu: 1000
    memory: 1000Gi
  disruption:
    consolidationPolicy: WhenEmptyOrUnderutilized
    consolidateAfter: 20m
**Versions**:
- Chart Version: v1.1.0

**Comment by leoryu** [comment]
The `startupTaints` in nodepool only works after node is registered which is added by karpenter core controller. The `karpenter.sh/unregistered` should be added by kubelet not the karpenter controller. You might check how the aws provider pr https://github.com/aws/karpenter-provider-aws/pull/6388/files.

**Comment by janani2019** [comment]
No, I’m exploring about karpenter for my eks cluster that hosts  GH runners. I’m using custom AMI. Could you please assist on how can I add the required taint karpenter.sh/unregistered via user data ?

**Comment by leoryu** [comment]
> No, I’m exploring about karpenter for my eks cluster that hosts GH runners. I’m using custom AMI. Could you please assist on how can I add the required taint karpenter.sh/unregistered via user data ?

If you are using karpenter-aws-provider, the provider will do these things. If the aws provider doesn't do this, I think you should add an issue in https://github.com/aws/karpenter-provider-aws not here.

**Comment by janani2019** [comment]
From the karpenter upgrade guide- “ Karpenter now adds a karpenter.sh/unregistered:NoExecute taint to nodes in injected UserData when using alias in AMISelectorTerms or non-Custom AMIFamily. When using amiFamily: Custom, users will need to add this taint into their UserData, where Karpenter will automatically remove it when provisioning nodes”. Would be helpful if you could elaborate on how taints can be added to user data .

**Comment by engedaam** [comment]
Here is more documentation on adding taints: https://karpenter.sh/docs/concepts/nodeclasses/#specuserdata

**Comment by geoffo-dev** [comment]
So I am having exactly the same issue - however I have added this as part of my UserData script:

```    
apiVersion: karpenter.k8s.aws/v1
    kind: EC2NodeClass
    metadata:
      name: default
    spec:
      amiFamily: Custom
      userData: |
        #!/bin/bash
        set -e
        B64_CLUSTER_CA=${module.eks.cluster_certificate_authority_data}
        API_SERVER_URL=${module.eks.cluster_endpoint}
        /etc/eks/bootstrap.sh ${module.eks.cluster_name} --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL \
          --kubelet-extra-args='--register-with-taints="karpenter.sh/unregistered:NoExecute"'
```

I can see the taint when the node is initialised, and then it is removed, however the nodeclaim still says `False` for ready with the same error:

```
  Conditions:
    Last Transition Time:  2024-11-15T17:10:23Z
    Message:
    Reason:                Initialized
    Status:                True
    Type:                  Initialized
    Last Transition Time:  2024-11-15T17:08:57Z
    Message:
    Reason:                Launched
    Status:                True
    Type:                  Launched
    Last Transition Time:  2024-11-15T17:10:23Z
    Message:               Registered=False
    Reason:                UnhealthyDependents
    Status:                False
    Type:                  Ready
    Last Transition Time:  2024-11-15T17:10:00Z
    Message:               Invariant violated, karpenter.sh/unregistered taint must be present on Karpenter-managed nodes
    Reason:                UnregisteredTaintNotFound
    Status:                False
    Type:                  Registered
  Image ID:                ami-01a816a693a50ab5c
  Provider ID:             aws:///eu-west-2a/i-0ec526c836ff8574f
Events:
  Type    Reason             Age    From       Message
  ----    ------             ----   ----       -------
  Normal  Launched           7m19s  karpenter  Status condition transitioned, Type: Launched, Status: Unknown -> True, Reason: Launched
  Normal  DisruptionBlocked  7m18s  karpenter  Cannot disrupt NodeClaim: state node doesn't contain both a node and a nodeclaim
  Normal  Ready              6m16s  karpenter  Status condition transitioned, Type: Ready, Status: Unknown -> False, Reason: UnhealthyDependents, Message: Initialized=Unknown, Registered
=False
  Normal  Registered         6m16s  karpenter  Status condition transitioned, Type: Registered, Status: Unknown -> False, Reason: UnregisteredTaintNotFound, Message: Invariant violated,
karpenter.sh/unregistered taint must be present on Karpenter-managed nodes
  Normal  Initialized        5m52s  karpenter  Status condition transitioned, Type: Initialized, Status: Unknown -> True, Reason: Initialized
  Normal  Unconsolidatable   5m48s  karpenter  NodePool "default" has consolidation disabled

```

I have been trying this with version 1.0.7 and 1.0.8...

**Comment by geoffo-dev** [comment]
have raised this under #1828 as I believe it to be seperate

**Comment by jonathan-innis** [comment]
Sounds like this is no longer a problem by looking at the linked issue (#1828)

/close
